### PR TITLE
Use test-tap's try and has functions

### PR DESCRIPTION
**Pull request by ingydotnet**
Based on #194

### CODE REVIEW

**Comment by okurz** [comment]
@mergify rebase

File: test/02-investigate.t [hunk 1]
```diff
@@ -64,53 +64,45 @@ openqa-cli() {
...
+has "$got" '1234' 'comment ID returned'
 
 # test syncing via investigation comment; we're second
-rc=0
-out=$(force=true sync_via_investigation_comment 32 32 2>&1) || rc=$?
+try force=true sync_via_investigation_comment 32 32
 is "$rc" 0 'skip with success if we do not own first investigation comment'
-like "$out" '' 'no output when skipping'
+# XXX What is this testing?
+like "$got" '' 'no output when skipping'
```
Comment (kalikiana) [inline test/02-investigate.t:105]: Should be https://github.com/os-autoinst/scripts/blob/master/openqa-investigate#L230

Comment (kalikiana) [reply test/02-investigate.t:105]: Generally there's a message that we're skipping. So this should ensure we don't in this case

File: test/01-label-known-issues.t [hunk 2]
```diff
@@ -73,22 +73,22 @@ testurl=https://openqa.opensuse.org/api/v1/jobs/2291399
...
+has "$out" 'Subject: Unreviewed issue (Group 24 openQA)' "send-email subject like expected"
+has "$out" 'From: openqa-label-known-issues <foo@bar>' "send-email from called like expected"
+has "$out" 'To: dummy@example.com.dummy' "send-email to like expected"
 like "$out" '8<.*Backend process died.*>8' 'Log excerpt in mail'
-like "$out" 'Content-Type: text/html' 'mail has text part'
-like "$out" 'Content-Type: text/plain' 'mail has HTML part'
-like "$out" '<li>Name: foo' 'mail contains job name'
+has "$out" 'Content-Type: text/html' 'mail has text part'
+has "$out" 'Content-Type: text/plain' 'mail has HTML part'
+has "$out" 'Name: foo' 'mail contains job name'
```
Comment (perlpunk) [inline test/01-label-known-issues.t:82]: why did you remove the `<li>`?

File: test/04-common.t [hunk 3]
```diff
@@ -15,14 +15,13 @@ failure() {
...
+try runcli success a b c
 is $rc 0 "runcli success"
-is "$output" "SUCCESS a b c" "runcli successful output"
+is "$got" "SUCCESS a b c" "runcli successful output"
 
-output=$(runcli failure a b c 2>&1) || rc=$?
+try runcli failure a b c
 is $rc 23 "runcli failure"
-like "$output", "test/04-common.t.*failure a b c.*oh noe" "runcli failure output"
+like "$got", "failure a b c.*oh noe" "runcli failure output"
```
Comment (perlpunk) [inline test/04-common.t:24]: I noticed you removed the `test/04-common.t.*` from the match.
This was actually making sure that the error message contained information where it happened:
```
openqa-cli (34 test/04-common.t): 
```
With this PR it says:
```
openqa-cli (168 /home/tina/openqa-devel/repos/scripts/bpan/.bpan/lib/test-tap.bash):
```
So I assume it needs to go one caller level deeper somewhere. Not sure if that's possible with `try`.

File: test/01-label-known-issues.t (hunk 2 above)
Comment (ingydotnet) [reply test/01-label-known-issues.t]: Fixed. See 5e8bef9 for details.

File: test/04-common.t (hunk 3 above)
Comment (ingydotnet) [reply test/04-common.t:24]: Right. There is stuff in test-tap-bash to adjust the call stack level on error messages.

I need to figure out a patch if you need this to have the file name.

BTW, just noticed the likely unintended comma in that line :)

I'll work on a fix for this.

Comment (ingydotnet) [reply test/04-common.t]: Fixed in da12258

File: test/02-investigate.t (hunk 1 above)
Comment (ingydotnet) [reply test/02-investigate.t:105]: @kalikiana the `like` testing function applies a regex to a string, so this is testing that the output _contains_ the empty string (regex), which would always be true.

I tried changing this to `is "$got" '' 'no output when skipping'` which I would expect is what you want, but the test fails.

I'll leave that for you to investigate. :)

**Review by okurz — COMMENTED** [review]
+1

**Review by okurz — APPROVED** [review]

File: _common [hunk 4]
```diff
@@ -36,7 +37,7 @@ runcli() {
     rc=$?
     set -e
     [[ "$rc" == 0 ]] && echo "$output" && return
-    warn "openqa-cli ($(caller)): (${args[*]}) rc: $rc >>>$output<<<"
+    warn "openqa-cli ($(caller "$runcli_caller_level")): (${args[*]}) rc: $rc >>>$output<<<"
```
Comment (perlpunk) [inline _common:40]: I was hoping for something more generic in the `try` function, e.g. in perl you can do:
```
some test_helper_function {
    local $Test::Builder::Level = $Test::Builder::Level + 1;
    ...
}
```
The error message including the caller we wouldn't just need for the `runcli` function, but I would expect the right caller level in every function we test.
OTOH the level just wrong in the tests, so I'd prefer to remove this commit again.

Comment (perlpunk) [reply _common:40]: Maybe it could be done by adding a wrapper around the `caller` function that uses such a `level` variable, and `try` will increase it.

Comment (ingydotnet) [reply _common]: That strategy worked in b4455c0.

**Comment by ingydotnet** [comment]
Squashed commits getting the warning caller to work in [test/04-common.t](https://github.com/os-autoinst/scripts/pull/196/files#diff-033dbd7ac4c2e9fab2c941a9379c1bc220198d43cff36d547a9b36a4abcb5909)

See 0804876

**Review by perlpunk — APPROVED** [review]
thanks!

**Comment by ingydotnet** [comment]
@perlpunk , you're welcome. Will rebase into #197 now.

**Review by kraih — APPROVED** [review]

File: test/02-investigate.t (hunk 1 above)
Comment (perlpunk) [reply test/02-investigate.t:105]: yeah, leave it like that and we will investigate it
